File: Karp-rabin/wordmatch.c
#include <cstdio>
#include <cmath>

#define PRIME 101
#define BASE 128

int hash(int n)
{
	return n%PRIME;
}

class RollingHash
{	
	private:
		int magic = hash(int(pow(BASE,  5)));
		char start;
		unsigned long long int sum=0;
	public:
		int hashvalue=0;
		RollingHash(char initstring[], int size)
		{
			for (int i = 0; i<5; ++i)
			{
				sum += hash(double(initstring[i]))*int(pow(BASE,i));
			}
			hashvalue = sum%PRIME;

		}

		void moveWindow(char front, char back)
		{
			hashvalue = ((hashvalue*BASE - int(front)*magic + int(back)) % PRIME );
		}				 
};

int main()
{
	char word[] = "hello";
	char test[] = "jjkj hello sfj jjjh hello";
	unsigned long long int sum = 0;
	for (int i = 0; word[i]!='\0'; ++i)
	{
		sum += (hash(int(word[i]))*int(pow(BASE,i)));
	}
	int hashinit = hash(sum);
	RollingHash R(test,5);
	for (int i = 0; test[i]!='\0' && i<int(sizeof(test)-5); ++i)
	{
		R.moveWindow(test[i],test[i+5]);
		if(hashinit == R.hashvalue)
		{
			printf("Found");
			break;
		}
	}
}

File: Karp-rabin/wordmatch.py
def dec(ch):
    return ord(ch) - ord('a') + 1


class RollingHash():
    def __init__(self, word, length):
        self.hash = 0
        self.length = length
        print(self.length)
        print(word[0:length])
        for i in range(0, length):
            self.hash += dec(word[i]) * 26 ** (length - 1 - i)

    def moveWindow(self, new, old):
        self.hash = (self.hash * 26) - (dec(old) *
                                        (26 ** (self.length))) + dec(new)


test = "ahello iam athul"
word = "hello"

wordR = RollingHash(word, len(word))
testR = RollingHash(test[0:len(word)], len(word))

for i in range(0, len(test) - len(word)):
    print(wordR.hash, testR.hash)
    if(wordR.hash == testR.hash):
        print("Found at ", i)
        break
    print(test[i + len(word)], test[i])
    testR.moveWindow(test[i + len(word)], test[i])
